### ISSUE TITLE
generics with a union bound aren't type checked

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#10948]
```py
from typing import Generic, TypeVar, Union


class Foo:
    ...


class Bar:
    ...


T = TypeVar("T", bound=Union[Foo, Bar])


class Baz(Generic[T]):
    def __init__(self, value: T):
        value.foo()  # no error
```
https://mypy-play.net/?mypy=latest&python=3.10&flags=show-column-numbers%2Cshow-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=37430e4ccdd3cc15aa76d318000f217b

**Comment by KotlinIsland** [comment]
```py
from typing import TypeVar, Union


T = TypeVar("T", bound=Union[int, str])

def foo(t: T) -> None:
    t.foo()  # no error

foo(1)
foo("1")
foo(True)  # no error
```
😬

**Comment by erictraut** [comment]
`foo(True)` shouldn't generate an error because `bool` is a subtype of `int`, so it matches the bound.

I think the other cases (`value.foo` in the first example and `t.foo` in the second example) should generate errors. Pyright generates errors in these cases.
